### ISSUE TITLE
Pyright incorrectly flags FAISS function with 'argument missing for parameter'

### ISSUE THREAD

**Issue by theahura** [microsoft/pyright#5219]
**Describe the bug**
FAISS is a library used for indexing and calculating distances over vectors. It is a python library with C bindings -- most of the heavy lifting is done in C. There is a function, `add`, that takes in a single parameter, `x`. This wraps an equivalent C function that takes in two parameters, `n` and `x`, where `n` is `len(x)`. 

Pyright seems to be using the C function signature instead of the python one. I get `Arguments missing for parameter "x"` when I would expect no error at all. 

I'm not sure if this is an issue with pyright or faiss, I crossposted an issue there as well: https://github.com/facebookresearch/faiss/issues/2891

**To Reproduce**
```py
import faiss
import numpy as np
emb = np.random.rand(800, 5)
idx = faiss.IndexFlatIP(emb.shape[-1])
idx.add(emb)  # works at runtime, but pyright fails with error: "Arguments missing for parameter 'x'"
idex.add(len(emb), emb)  # pyright is happy, but this fails at runtime because the wrong number of args are given
```

**Expected behavior**
Pyright would compile without issue

**Screenshots or Code**
See above

**VS Code extension or command-line**
```
pyright test.py
```

**Additional context**
NA

**Comment by erictraut** [comment]
Pyright doesn't know anything about C functions, so that wouldn't explain the behavior you're seeing.

The `faiss` library (I installed using `pip install faiss-cpu`) is not a typed library. That means pyright will either treat all imported symbols as unknown (if you have `useLibraryCodeForTypes` set to false) or attempt to infer type information from the available Python source code shipped with the library. I presume that you have `useLibraryCodeForTypes` enabled, which is the default.

The type of `idx` in your code sample above is evaluated as `IndexFlatIP`, which is a class defined in the source file `faiss/swigfaiss.py`. Here's the relevant code:

```python
class IndexFlatIP(IndexFlat):
    ...

class IndexFlat(IndexFlatCodes):
    ...

class IndexFlatCodes(Index):
    ...
    def add(self, n, x):
        r""" default add uses sa_encode"""
        return _swigfaiss.IndexFlatCodes_add(self, n, x)
```

As you can see, the `add` method requires two arguments.

It appears that the library is dynamically monkey patching the `add` method with a replacement called `replacement_add`, and the replacement method has a different signature than the original. I don't know the reason behind this, but this won't work with static type checkers or language servers. I recommend filing an issue with the maintainers of the `faiss` library.

Your options include:
1. Suppress the error using a `# type: ignore` comment.
2. Find (or write your own) type stubs for this library that provide accurate type information.
3. Disable `useLibraryCodeForTypes`. This will eliminate the error but will also eliminate all completion suggestions for the library.
